File: src/render/transitions/shared.ts
import path from 'node:path';
import fs from 'node:fs';
import { execFileSync } from 'node:child_process';
import type { TemplateInfo } from '../../registry/index.js';
import type { QualityPreset, ResolutionPreset } from '../../presets.js';
import { updateProgress, type RenderTask } from '../task.js';
import type { ProbedVideoInfo, RenderRequest, VideoClip } from '../types.js';

export interface TransitionHandlerArgs {
  rootDir: string;
  ffmpegPath: string;
  ffprobePath: string;
  request: RenderRequest;
  videoInfo: Map<string, ProbedVideoInfo>;
  qualPreset: QualityPreset;
  resPreset: ResolutionPreset;
  task: RenderTask;
  outDir: string;
  outFile: string;
}

export interface TransitionHandlerResult {
  cleanup: () => void;
  outputPath: string;
}

export function collectClips(
  variables: Record<string, unknown>,
  templateInfo: TemplateInfo,
  videoInfo: Map<string, ProbedVideoInfo>,
): VideoClip[] {
  for (const [key, def] of Object.entries(templateInfo.manifest.variables)) {
    if (def.type !== 'video_list') continue;
    const list = variables[key];
    if (!Array.isArray(list)) continue;
    const durations = (variables[`${key}_source_durations`] as number[] | undefined) ?? [];
    return list.map((src, index) => ({
      key: `clip_${index + 1}`,
      src: src as string,
      duration: durations[index] ?? 0,
    }));
  }

  const clips: VideoClip[] = [];
  for (const [key, def] of Object.entries(templateInfo.manifest.variables)) {
    if (def.type !== 'video') continue;
    const src = variables[key];
    if (typeof src !== 'string' || !src.trim()) continue;
    const info = videoInfo.get(key);
    clips.push({ key, src, duration: info?.duration ?? 0 });
  }
  return clips;
}

export function buildNormalizeVideoFilter(resPreset: ResolutionPreset): string {
  return [
    `fps=${resPreset.fps}`,
    `scale=${resPreset.width}:${resPreset.height}:force_original_aspect_ratio=decrease:flags=lanczos`,
    `pad=${resPreset.width}:${resPreset.height}:(ow-iw)/2:(oh-ih)/2:color=black`,
    'setsar=1',
    'format=yuv420p',
  ].join(',');
}

export function normalizeClips(
  rootDir: string,
  ffmpegPath: string,
  clips: VideoClip[],
  qualPreset: QualityPreset,
  resPreset: ResolutionPreset,
): {
  clips: VideoClip[];
  cleanup: () => void;
} {
  const tempDir = path.join(rootDir, 'temp');
  fs.mkdirSync(tempDir, { recursive: true });

  const normalizeFilter = buildNormalizeVideoFilter(resPreset);
  const sessionId = `${Date.now().toString(36)}_${Math.random().toString(36).slice(2, 8)}`;
  const tempFiles: string[] = [];

  const cleanup = () => {
    for (const file of tempFiles) {
      try {
        fs.unlinkSync(file);
      } catch {
        // ignore temp cleanup errors
      }
    }
  };

  try {
    console.log(
      `  [normalize] all input videos -> ${resPreset.width}x${resPreset.height} ${resPreset.fps}fps (lanczos)`,
    );

    const normalizedClips = clips.map((clip, index) => {
      const tempFile = path.join(tempDir, `normalized_${sessionId}_${index}.mp4`);
      tempFiles.push(tempFile);

      execFileSync(
        ffmpegPath,
        [
          '-i', clip.src,
          '-vf', normalizeFilter,
          '-c:v', 'libx264',
          '-preset', qualPreset.ffmpegPreset,
          '-crf', String(qualPreset.crf),
          '-pix_fmt', 'yuv420p',
          '-an',
          '-y', tempFile,
        ],
        { timeout: 600_000 },
      );

      return {
        ...clip,
        src: tempFile,
      };
    });

    return {
      clips: normalizedClips,
      cleanup,
    };
  } catch (error) {
    cleanup();
    throw error;
  }
}

export function ffmpegTrimConcat(
  rootDir: string,
  ffmpegPath: string,
  ffprobePath: string,
  clips: VideoClip[],
  outputPath: string,
  qualPreset: QualityPreset,
  task: RenderTask,
  trimStart: number,
): void {
  const tempDir = path.join(rootDir, 'temp');
  fs.mkdirSync(tempDir, { recursive: true });

  const sessionId = Date.now().toString(36);
  const tempFiles: string[] = [];

  try {
    void ffprobePath;

    console.log(`\n[2/3] FFmpeg 裁剪 ${clips.length} 个片段...`);

    for (let i = 0; i < clips.length; i++) {
      const clip = clips[i];
      const tempFile = path.join(tempDir, `trim_${sessionId}_${i}.mp4`);
      tempFiles.push(tempFile);

      const args: string[] = [
        '-ss', String(trimStart),
        '-i', clip.src,
        '-c:v', 'libx264',
        '-preset', qualPreset.ffmpegPreset,
        '-crf', String(qualPreset.crf),
        '-an',
        '-y', tempFile,
      ];

      console.log(`  [${i + 1}/${clips.length}] 裁剪 ${path.basename(clip.src)}（跳过前 ${trimStart}s）`);
      execFileSync(ffmpegPath, args, { timeout: 300_000 });

      updateProgress(task, (i + 1) / (clips.length + 1));
    }

    const listFile = path.join(tempDir, `concat_${sessionId}.txt`);
    const listContent = tempFiles
      .map((file) => `file '${file.replace(/\\/g, '/')}'`)
      .join('\n');
    fs.writeFileSync(listFile, listContent, 'utf-8');

    console.log(`  拼接 ${clips.length} 个片段 -> ${path.basename(outputPath)}`);
    execFileSync(
      ffmpegPath,
      [
        '-f', 'concat',
        '-safe', '0',
        '-i', listFile,
        '-c', 'copy',
        '-y', outputPath,
      ],
      { timeout: 600_000 },
    );

    try {
      fs.unlinkSync(listFile);
    } catch {
      // ignore temp cleanup errors
    }
  } finally {
    for (const file of tempFiles) {
      try {
        fs.unlinkSync(file);
      } catch {
        // ignore temp cleanup errors
      }
    }
  }
}

File: videocut/render/transitions/shared.py
from __future__ import annotations

import os
import subprocess
from pathlib import Path
from typing import Callable

from videocut.errors import RenderError
from videocut.ffmpeg_config import FFmpegVideoSettings
from videocut.log import get_logger
from videocut.presets import QualityPreset, ResolutionPreset
from videocut.render.types import VideoClip

logger = get_logger(__name__)
_FFMPEG_ERROR_TAIL_CHARS = 4000


def _tail_text(value: str | None) -> str:
    if not value:
        return ""
    text = value.strip()
    if len(text) <= _FFMPEG_ERROR_TAIL_CHARS:
        return text
    return "...\n" + text[-_FFMPEG_ERROR_TAIL_CHARS:]


def _format_ffmpeg_failure(label: str, returncode: int, stderr: str | None) -> str:
    detail = f"{label}: FFmpeg exited with code {returncode}."
    stderr_tail = _tail_text(stderr)
    if stderr_tail:
        detail += f"\n  stderr tail:\n{stderr_tail}"
    return detail


def build_normalize_video_filter(res_preset: ResolutionPreset) -> str:
    return ",".join(
        [
            f"fps={res_preset.fps}",
            (
                f"scale={res_preset.width}:{res_preset.height}:"
                "force_original_aspect_ratio=decrease:flags=lanczos"
            ),
            f"pad={res_preset.width}:{res_preset.height}:(ow-iw)/2:(oh-ih)/2:color=black",
            "setsar=1",
            "format=yuv420p",
        ]
    )


def run_ffmpeg_checked(args: list[str], timeout: int, *, label: str = "ffmpeg") -> None:
    try:
        result = subprocess.run(
            args,
            check=False,
            timeout=timeout,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            encoding="utf-8",
            errors="replace",
        )
    except subprocess.CalledProcessError as exc:
        raise RenderError(_format_ffmpeg_failure(label, exc.returncode, exc.stderr)) from exc
    except subprocess.TimeoutExpired as exc:
        stderr_tail = _tail_text(exc.stderr.decode("utf-8", errors="replace") if isinstance(exc.stderr, bytes) else exc.stderr)
        detail = f"{label}: FFmpeg timed out after {timeout} seconds."
        if stderr_tail:
            detail += f"\n  stderr tail:\n{stderr_tail}"
        raise RenderError(detail) from exc
    if result.returncode != 0:
        raise RenderError(_format_ffmpeg_failure(label, result.returncode, result.stderr))


def _run_ffmpeg(args: list[str], timeout: int) -> None:
    run_ffmpeg_checked(args, timeout)


def normalize_clips(
    root_dir: Path,
    ffmpeg_path: str,
    clips: list[VideoClip],
    qual_preset: QualityPreset,
    res_preset: ResolutionPreset,
    video_settings: FFmpegVideoSettings,
) -> tuple[list[VideoClip], Callable[[], None]]:
    temp_dir = root_dir / "temp"
    temp_dir.mkdir(parents=True, exist_ok=True)
    normalize_filter = build_normalize_video_filter(res_preset)
    session_id = f"{os.getpid():x}_{Path.cwd().stat().st_mtime_ns:x}"
    temp_files: list[Path] = []

    def cleanup() -> None:
        for file_path in temp_files:
            try:
                file_path.unlink()
            except FileNotFoundError:
                pass

    try:
        logger.info(
            "  [normalize] all input videos -> %dx%d %dfps (lanczos)",
            res_preset.width, res_preset.height, res_preset.fps,
        )
        normalized: list[VideoClip] = []
        for index, clip in enumerate(clips):
            temp_file = temp_dir / f"normalized_{session_id}_{index}.mp4"
            temp_files.append(temp_file)
            run_ffmpeg_checked(
                [
                    ffmpeg_path,
                    *video_settings.input_args(),
                    "-i",
                    clip.src,
                    "-vf",
                    normalize_filter,
                    *video_settings.output_args(qual_preset),
                    "-an",
                    "-y",
                    str(temp_file),
                ],
                timeout=600,
                label=f"normalize clip {index + 1} ({Path(clip.src).name})",
            )
            normalized.append(VideoClip(key=clip.key, src=str(temp_file), duration=clip.duration))
        return normalized, cleanup
    except Exception:
        cleanup()
        raise
